// @waves/waves-browser-bus/dist/bus/Bus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
var EventType;
(function (EventType) {
    EventType[EventType["Event"] = 0] = "Event";
    EventType[EventType["Action"] = 1] = "Action";
    EventType[EventType["Response"] = 2] = "Response";
})(EventType = exports.EventType || (exports.EventType = {}));
var ResponseStatus;
(function (ResponseStatus) {
    ResponseStatus[ResponseStatus["Success"] = 0] = "Success";
    ResponseStatus[ResponseStatus["Error"] = 1] = "Error";
})(ResponseStatus = exports.ResponseStatus || (exports.ResponseStatus = {}));
var Bus = /** @class */ (function () {
    function Bus(adapter, defaultTimeout) {
        var _this = this;
        this.id = utils_1.uniqueId('bus');
        this._timeout = defaultTimeout || 5000;
        this._adapter = adapter;
        this._adapter.addListener(function (data) { return _this._onMessage(data); });
        this._eventHandlers = Object.create(null);
        this._activeRequestHash = Object.create(null);
        this._requestHandlers = Object.create(null);
        utils_1.console.info("Create Bus with id \"" + this.id + "\"");
    }
    Bus.prototype.dispatchEvent = function (name, data) {
        this._adapter.send(Bus._createEvent(name, data));
        utils_1.console.info("Dispatch event \"" + name + "\"", data);
        return this;
    };
    Bus.prototype.request = function (name, data, timeout) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var id = utils_1.uniqueId(_this.id + "-action");
            var wait = timeout || _this._timeout;
            var timer;
            if ((timeout || _this._timeout) !== -1) {
                timer = setTimeout(function () {
                    delete _this._activeRequestHash[id];
                    var error = new Error("Timeout error for request with name \"" + name + "\" and timeout " + wait + "!");
                    utils_1.console.error(error);
                    reject(error);
                }, wait);
            }
            var cancelTimeout = function () {
                if (timer) {
                    clearTimeout(timer);
                }
            };
            _this._activeRequestHash[id] = {
                reject: function (error) {
                    cancelTimeout();
                    utils_1.console.error("Error request with name \"" + name + "\"", error);
                    reject(error);
                },
                resolve: function (data) {
                    cancelTimeout();
                    utils_1.console.info("Request with name \"" + name + "\" success resolved!", data);
                    resolve(data);
                }
            };
            _this._adapter.send({ id: id, type: 1 /* Action */, name: name, data: data });
            utils_1.console.info("Request with name \"" + name + "\"", data);
        });
    };
    Bus.prototype.on = function (name, handler, context) {
        return this._addEventHandler(name, handler, context, false);
    };
    Bus.prototype.once = function (name, handler, context) {
        return this._addEventHandler(name, handler, context, true);
    };
    Bus.prototype.off = function (name, handler) {
        var _this = this;
        if (!name) {
            Object.keys(this._eventHandlers).forEach(function (name) { return _this.off(name, handler); });
            return this;
        }
        if (!this._eventHandlers[name]) {
            return this;
        }
        if (!handler) {
            this._eventHandlers[name].slice().forEach(function (info) {
                _this.off(name, info.handler);
            });
            return this;
        }
        this._eventHandlers[name] = this._eventHandlers[name].filter(function (info) { return info.handler !== handler; });
        if (!this._eventHandlers[name].length) {
            delete this._eventHandlers[name];
        }
        return this;
    };
    Bus.prototype.registerRequestHandler = function (name, handler) {
        if (this._requestHandlers[name]) {
            throw new Error('Duplicate request handler!');
        }
        this._requestHandlers[name] = handler;
        return this;
    };
    Bus.prototype.unregisterHandler = function (name) {
        if (this._requestHandlers[name]) {
            delete this._requestHandlers[name];
        }
        return this;
    };
    Bus.prototype.changeAdapter = function (adapter) {
        var _this = this;
        var bus = new Bus(adapter, this._timeout);
        Object.keys(this._eventHandlers).forEach(function (name) {
            _this._eventHandlers[name].forEach(function (info) {
                if (info.once) {
                    bus.once(name, info.handler, info.context);
                }
                else {
                    bus.on(name, info.handler, info.context);
                }
            });
        });
        Object.keys(this._requestHandlers).forEach(function (name) {
            bus.registerRequestHandler(name, _this._requestHandlers[name]);
        });
        return bus;
    };
    Bus.prototype.destroy = function () {
        utils_1.console.info('Destroy Bus');
        this.off();
        this._adapter.destroy();
    };
    Bus.prototype._addEventHandler = function (name, handler, context, once) {
        if (!this._eventHandlers[name]) {
            this._eventHandlers[name] = [];
        }
        this._eventHandlers[name].push({ handler: handler, once: once, context: context });
        return this;
    };
    Bus.prototype._onMessage = function (message) {
        switch (message.type) {
            case 0 /* Event */:
                utils_1.console.info("Has event with name \"" + message.name + "\"", message.data);
                this._fireEvent(message.name, message.data);
                break;
            case 1 /* Action */:
                utils_1.console.info("Start action with id \"" + message.id + "\" and name \"" + message.name + "\"", message.data);
                this._createResponse(message);
                break;
            case 2 /* Response */:
                utils_1.console.info("Start response with name \"" + message.id + "\" and status \"" + message.status + "\"", message.content);
                this._fireEndAction(message);
                break;
        }
    };
    Bus.prototype._createResponse = function (message) {
        var _this = this;
        var sendError = function (error) {
            utils_1.console.error(error);
            _this._adapter.send({
                id: message.id,
                type: 2 /* Response */,
                status: 1 /* Error */,
                content: String(error)
            });
        };
        if (!this._requestHandlers[message.name]) {
            sendError(new Error("Has no handler for \"" + message.name + "\" action!"));
            return void 0;
        }
        try {
            var result = this._requestHandlers[message.name](message.data);
            if (Bus._isPromise(result)) {
                result.then(function (data) {
                    _this._adapter.send({
                        id: message.id,
                        type: 2 /* Response */,
                        status: 0 /* Success */,
                        content: data
                    });
                }, sendError);
            }
            else {
                this._adapter.send({
                    id: message.id,
                    type: 2 /* Response */,
                    status: 0 /* Success */,
                    content: result
                });
            }
        }
        catch (e) {
            sendError(e);
        }
    };
    Bus.prototype._fireEndAction = function (message) {
        if (this._activeRequestHash[message.id]) {
            switch (message.status) {
                case 1 /* Error */:
                    this._activeRequestHash[message.id].reject(message.content);
                    break;
                case 0 /* Success */:
                    this._activeRequestHash[message.id].resolve(message.content);
                    break;
            }
            delete this._activeRequestHash[message.id];
        }
    };
    Bus.prototype._fireEvent = function (name, value) {
        if (!this._eventHandlers[name]) {
            return void 0;
        }
        this._eventHandlers[name] = this._eventHandlers[name]
            .slice()
            .filter(function (handlerInfo) {
            try {
                handlerInfo.handler.call(handlerInfo.context, value);
            }
            catch (e) {
                utils_1.console.warn(e);
            }
            return !handlerInfo.once;
        });
        if (!this._eventHandlers[name].length) {
            delete this._eventHandlers[name];
        }
    };
    Bus._createEvent = function (eventName, data) {
        return {
            type: 0 /* Event */,
            name: eventName,
            data: data
        };
    };
    Bus._isPromise = function (some) {
        return some && some.then && typeof some.then === 'function';
    };
    return Bus;
}());
exports.Bus = Bus;
