// ts-utils/src/Signal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("./utils");
var Signal = (function () {
    function Signal() {
        this.cid = utils_1.uniqueId('signal');
        this._handlers = [];
    }
    Signal.prototype.on = function (handler, context, receiver) {
        this._handlers.push({
            isOnce: false,
            handler: handler,
            context: context,
            receiver: receiver
        });
    };
    Signal.prototype.once = function (handler, context, receiver) {
        this._handlers.push({
            isOnce: true,
            handler: handler,
            context: context,
            receiver: receiver
        });
    };
    Signal.prototype.off = function (handler) {
        var _this = this;
        if (handler) {
            for (var i = this._handlers.length; i--;) {
                var handlerData = this._handlers[i];
                if (handlerData.handler === handler) {
                    this._handlers.splice(i, 1);
                    if (handlerData.receiver) {
                        handlerData.receiver.stopReceive(this, handler);
                    }
                }
            }
        }
        else {
            var handlers = this._handlers.slice();
            this._handlers = [];
            handlers.forEach(function (data) {
                if (data.receiver) {
                    data.receiver.stopReceive(_this, data.handler);
                }
            });
        }
    };
    Signal.prototype.dispatch = function (some) {
        var _this = this;
        this._handlers.slice().forEach(function (handlerData) {
            if (handlerData.isOnce) {
                _this.off(handlerData.handler);
            }
            handlerData.handler.call(handlerData.context, some);
        });
    };
    return Signal;
}());
exports.Signal = Signal;
