// ts-utils/src/tree/BaseTree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
var BaseTree = (function () {
    function BaseTree(data, parent, options) {
        var _this = this;
        this.children = [];
        this.parent = parent;
        this.id = data.id;
        this.data = utils_1.cloneDeep(data.data);
        this.ChildConstructor = options && options.Child || BaseTree;
        this.getRoot().registerChild(this);
        if (data.children) {
            data.children.forEach(function (item) {
                _this.children.push(new _this.ChildConstructor(item, _this, options));
            });
        }
    }
    BaseTree.prototype.getChildren = function () {
        return this.children || [];
    };
    BaseTree.prototype.toArray = function () {
        var result = [this];
        if (this.children) {
            this.children.forEach(function (item) {
                result.push.apply(result, item.toArray());
            });
        }
        return result;
    };
    BaseTree.prototype.getData = function () {
        return this.data || Object.create(null);
    };
    BaseTree.prototype.set = function (key, value) {
        this.data[key] = value;
    };
    BaseTree.prototype.get = function (key) {
        return this.getData()[key];
    };
    BaseTree.prototype.getExtended = function (key) {
        var result = this.get(key);
        return result == null ? this.parent.getExtended(key) : result;
    };
    BaseTree.prototype.getParent = function () {
        return this.parent;
    };
    BaseTree.prototype.getRoot = function () {
        return this.parent.getRoot();
    };
    return BaseTree;
}());
exports.BaseTree = BaseTree;
