// @waves/money-like-to-node/dist/toNodeEntities/exchange.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var factory_1 = require("../core/factory");
var general_1 = require("./general");
var utils_1 = require("../utils");
var getAssetPair = factory_1.factory({
    amountAsset: utils_1.pipe(utils_1.prop('amount'), utils_1.getAssetId),
    priceAsset: utils_1.pipe(utils_1.prop('price'), utils_1.getAssetId)
});
exports.remapOrder = factory_1.factory({
    version: utils_1.prop('version'),
    matcherPublicKey: utils_1.prop('matcherPublicKey'),
    orderType: utils_1.prop('orderType'),
    timestamp: utils_1.prop('timestamp'),
    expiration: utils_1.prop('expiration'),
    senderPublicKey: utils_1.prop('senderPublicKey'),
    proofs: utils_1.prop('proofs'),
    price: utils_1.pipe(utils_1.prop('price'), utils_1.getCoins),
    amount: utils_1.pipe(utils_1.prop('amount'), utils_1.getCoins),
    matcherFee: utils_1.pipe(utils_1.prop('matcherFee'), utils_1.getCoins),
    matcherFeeAssetId: utils_1.pipe(utils_1.prop('matcherFee'), utils_1.getAssetId),
    assetPair: getAssetPair
});
exports.exchange = factory_1.factory(__assign({}, general_1.getDefaultTransform(), { buyOrder: utils_1.pipe(utils_1.prop('buyOrder'), exports.remapOrder), sellOrder: utils_1.pipe(utils_1.prop('sellOrder'), exports.remapOrder), price: utils_1.pipe(utils_1.prop('price'), utils_1.getCoins), amount: utils_1.pipe(utils_1.prop('amount'), utils_1.getCoins), buyMatcherFee: utils_1.pipe(utils_1.prop('buyMatcherFee'), utils_1.getCoins), sellMatcherFee: utils_1.pipe(utils_1.prop('sellMatcherFee'), utils_1.getCoins) }));
