// @waves/oracle-data/dist/parse/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var response_1 = require("../response");
var constants_1 = require("../constants");
var schemas_1 = require("../schemas");
function parseOracleData(hash) {
    try {
        var version = response_1.getFieldValue(hash, constants_1.DATA_PROVIDER_KEYS.VERSION, constants_1.DATA_ENTRY_TYPES.INTEGER);
        return schemas_1.DATA_PROVIDER_VERSION_MAP[version](hash);
    }
    catch (e) {
        return {
            status: constants_1.RESPONSE_STATUSES.ERROR,
            content: {},
            errors: [{
                    path: 'version',
                    error: e
                }]
        };
    }
}
exports.parseOracleData = parseOracleData;
function parseAssetData(hash) {
    return Object.keys(hash)
        .map(response_1.getAssetIdFromKey)
        .filter(response_1.isString)
        .map(function (id) {
        try {
            var version = response_1.getFieldValue(hash, getDataName("version_<ASSET_ID>", id), constants_1.DATA_ENTRY_TYPES.INTEGER);
            var status_1 = response_1.getFieldValue(hash, getDataName("status_<ASSET_ID>", id), constants_1.DATA_ENTRY_TYPES.INTEGER);
            return schemas_1.ASSETS_VERSION_MAP[version](id, status_1)(hash);
        }
        catch (e) {
            return {
                status: constants_1.RESPONSE_STATUSES.ERROR,
                content: { id: id },
                errors: [{
                        path: 'version',
                        error: e
                    }]
            };
        }
    });
}
exports.parseAssetData = parseAssetData;
function getDataName(name, id) {
    return name.replace(constants_1.PATTERNS.ASSET_ID, "<" + id + ">");
}
