// ts-utils/src/Signal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("./utils");
var Signal = (function () {
    function Signal() {
        this.cid = utils_1.uniqueId('signal');
        this._handlers = [];
    }
    Signal.prototype.on = function (handler, context, receiver) {
        this._handlers.push({
            isOnce: false,
            handler: handler,
            context: context,
            receiver: receiver
        });
    };
    Signal.prototype.once = function (handler, context, receiver) {
        this._handlers.push({
            isOnce: true,
            handler: handler,
            context: context,
            receiver: receiver
        });
    };
    Signal.prototype.off = function (handler, receiver) {
        for (var i = this._handlers.length; i--;) {
            var handlerData = this._handlers[i];
            if (!handlerData) {
                continue;
            }
            if (handlerData.handler === handler) {
                if (handlerData.receiver) {
                    if (receiver) {
                        if (handlerData.receiver === receiver) {
                            this._handlers.splice(i, 1);
                            handlerData.receiver.stopReceive(this, handler);
                        }
                    }
                    else {
                        throw new Error('Can\'t remove this handler without receiver!');
                    }
                }
                else {
                    this._handlers.splice(i, 1);
                }
            }
        }
    };
    Signal.prototype.dispatch = function (some) {
        var _this = this;
        this._handlers.slice().forEach(function (handlerData) {
            if (handlerData.isOnce) {
                _this.off(handlerData.handler, handlerData.receiver);
            }
            handlerData.handler.call(handlerData.context, some);
        });
    };
    return Signal;
}());
exports.Signal = Signal;
