// @waves/money-like-to-node/dist/validators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBase64 = function (value) {
    var regExp = /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/;
    return regExp.test(value);
};
function isString(value) {
    return typeof value === 'string';
}
exports.isString = isString;
exports.required = function (value) {
    return value != null;
};
exports.charsInDictionary = function (dictionary) { return function (value) { return value.split('').every(function (char) { return dictionary.includes(char); }); }; };
exports.createValidator = function (validateFunction, message) { return function (data) {
    if (validateFunction(data)) {
        return null;
    }
    else {
        return message;
    }
}; };
exports.requiredValidator = function (property) { return exports.createValidator(exports.required, "Property \"" + property + "\" is required!"); };
function validate() {
    var validators = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        validators[_i] = arguments[_i];
    }
    return function (data) {
        var errors = validators.reduce(function (errors, validator) {
            var message = validator(data);
            if (message) {
                errors.push(message);
            }
            return errors;
        }, []);
        if (errors.length) {
            throw new Error("Validation error! Details: " + JSON.stringify(errors, null, 4));
        }
        return data;
    };
}
exports.validate = validate;
