// @waves/signature-adapter/dist/prepareTx/fieldTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fieldFactory = function (type, optionalData) { return function (fromField, toField, processor, optional) {
    if (toField === void 0) { toField = fromField; }
    if (processor === void 0) { processor = null; }
    if (optional === void 0) { optional = false; }
    return ({
        name: fromField,
        field: toField || fromField,
        optional: optional,
        processor: processor,
        type: type,
        optionalData: optionalData,
    });
}; };
exports.string = function (data) { return fieldFactory('string'); };
exports.asset = function (data) { return fieldFactory('assetId'); };
exports.assetName = function (data) { return fieldFactory('assetName'); };
exports.assetDescription = function (data) { return fieldFactory('assetDescription'); };
exports.precision = function (data) { return fieldFactory('precision'); };
exports.number = function (data) { return fieldFactory('number'); };
exports.address = function (data) { return fieldFactory('address', data); };
exports.aliasName = function (data) { return fieldFactory('aliasName', data); };
exports.aliasOrAddress = function (data) { return fieldFactory('aliasOrAddress', data); };
exports.money = function (data) { return fieldFactory('money'); };
exports.numberLike = function (data) { return fieldFactory('numberLike'); };
exports.attachment = function (data) { return fieldFactory('attachment'); };
exports.httpsUrl = function (data) { return fieldFactory('httpsUrl'); };
exports.timestamp = function (data) { return fieldFactory('timestamp'); };
exports.orderType = function (data) { return fieldFactory('orderType'); };
exports.fromData = function (data) { return fieldFactory('fromData'); };
exports.boolean = function (data) { return fieldFactory('boolean'); };
exports.transfers = function (data) { return fieldFactory('transfers', data); };
exports.data = function (data) { return fieldFactory('data'); };
exports.script = function (data) { return fieldFactory('script'); };
exports.asset_script = function (data) { return fieldFactory('asset_script'); };
exports.required = function (data) { return fieldFactory('required'); };
exports.call = function (data) { return fieldFactory('call'); };
exports.payment = function (data) { return fieldFactory('payment'); };
