// ts-api-validator/dist/ts-api-validator.partial.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.tsApiValidator = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./src/Schema"));
__export(require("./src/config"));
__export(require("./src/BooleanPart"));
__export(require("./src/StringPart"));
__export(require("./src/DatePart"));
__export(require("./src/StringDatePart"));
__export(require("./src/NumberPart"));
__export(require("./src/ObjectPart"));
__export(require("./src/ArrayPart"));
__export(require("./src/BasePart"));

},{"./src/ArrayPart":2,"./src/BasePart":3,"./src/BooleanPart":4,"./src/DatePart":5,"./src/NumberPart":6,"./src/ObjectPart":7,"./src/Schema":8,"./src/StringDatePart":9,"./src/StringPart":10,"./src/config":11}],2:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var BasePart_1 = require("./BasePart");
var ts_utils_1 = require("ts-utils");
var ArrayPart = (function (_super) {
    __extends(ArrayPart, _super);
    function ArrayPart(config, path) {
        var _this = _super.call(this, config, path) || this;
        var Component = _this.options.content.type;
        _this._child = new Component(_this.options.content);
        return _this;
    }
    ArrayPart.prototype.process = function (data, rootList) {
        var _this = this;
        return _super.prototype.process.call(this, data, rootList).then(function (value) {
            if (value && ts_utils_1.isArray(value)) {
                return Promise.all(value.map(function (item) { return _this._child.process(item, rootList.concat(data)); }));
            }
            else {
                return value;
            }
        });
    };
    ArrayPart.prototype.getValue = function (data) {
        if (ts_utils_1.isArray(data)) {
            return data;
        }
        else {
            return null;
        }
    };
    return ArrayPart;
}(BasePart_1.BasePart));
exports.ArrayPart = ArrayPart;

},{"./BasePart":3,"ts-utils":undefined}],3:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ts_utils_1 = require("ts-utils");
var BasePart = (function () {
    function BasePart(options, path) {
        this.options = options;
        this.path = path;
        if (this.options.isEmpty) {
            this.isEmpty = this.options.isEmpty;
        }
        if (this.options.isValid) {
            this.isValid = this.options.isValid;
        }
        if (this.options.required && ('defaultValue' in this.options)) {
            throw new Error('Wrong params! Conflict options "required" and defaultValue');
        }
    }
    BasePart.prototype.process = function (data, roots) {
        var _this = this;
        var path = this.getPath();
        var result = this.getValue(this.getDataByPath(data, path), roots);
        return BasePart.toPromise(result).then(function (value) {
            var isEmpty = _this.isEmpty(value);
            var isValid = _this.isValid(value);
            var type = _this.options.type.name || _this.options.type.prototype.constructor.name;
            if (_this.options.required) {
                if (isEmpty) {
                    throw new Error("Required field type \"" + type + "\" \"" + path + "\" is empty!");
                }
            }
            if (('defaultValue' in _this.options) && isEmpty) {
                value = _this.options.defaultValue;
            }
            else {
                if (!isValid) {
                    throw new Error("Field \"" + path + "\" is invalid!");
                }
            }
            return value;
        });
    };
    BasePart.prototype.getPath = function () {
        return this.options.path === null ? null : this.options.path || this.path;
    };
    BasePart.prototype.isEmpty = function (data) {
        return data == null;
    };
    BasePart.prototype.isValid = function (data) {
        return true;
    };
    BasePart.prototype.getDataByPath = function (data, path) {
        if (this.options.parseValue) {
            if (path) {
                return this.options.parseValue(ts_utils_1.get(data, path));
            }
            else {
                return this.options.parseValue(data);
            }
        }
        else if (path != null) {
            return ts_utils_1.get(data, path);
        }
        else {
            return data;
        }
    };
    BasePart.isPromise = function (some) {
        return some && some.then && typeof some.then === 'function';
    };
    BasePart.toPromise = function (some) {
        return BasePart.isPromise(some) ? some : Promise.resolve(some);
    };
    return BasePart;
}());
exports.BasePart = BasePart;

},{"ts-utils":undefined}],4:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var BasePart_1 = require("./BasePart");
var BooleanPart = (function (_super) {
    __extends(BooleanPart, _super);
    function BooleanPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BooleanPart.prototype.getValue = function (data) {
        switch (typeof data) {
            case 'boolean':
                return data;
            case 'string':
            case 'number':
                return Boolean(data);
            default:
                return null;
        }
    };
    return BooleanPart;
}(BasePart_1.BasePart));
exports.BooleanPart = BooleanPart;

},{"./BasePart":3}],5:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var BasePart_1 = require("./BasePart");
var DatePart = (function (_super) {
    __extends(DatePart, _super);
    function DatePart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DatePart.prototype.getValue = function (data) {
        if (data instanceof Date) {
            return data;
        }
        if (typeof data === 'number') {
            return new Date(data);
        }
        return null;
    };
    return DatePart;
}(BasePart_1.BasePart));
exports.DatePart = DatePart;

},{"./BasePart":3}],6:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var BasePart_1 = require("./BasePart");
var NumberPart = (function (_super) {
    __extends(NumberPart, _super);
    function NumberPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NumberPart.prototype.getValue = function (data) {
        switch (typeof data) {
            case 'number':
                return data;
            case 'string':
                return Number(data);
            default:
                return null;
        }
    };
    NumberPart.prototype.isEmpty = function (data) {
        return data == null || isNaN(data);
    };
    return NumberPart;
}(BasePart_1.BasePart));
exports.NumberPart = NumberPart;

},{"./BasePart":3}],7:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var BasePart_1 = require("./BasePart");
var ts_utils_1 = require("ts-utils");
var ObjectPart = (function (_super) {
    __extends(ObjectPart, _super);
    function ObjectPart(config, path) {
        var _this = _super.call(this, config, path) || this;
        var myPath = _this.getPath();
        _this._childHash = Object.create(null);
        ts_utils_1.each(_this.options.content, function (config, key) {
            var Component = config.type;
            var localPath = path == null ? String(key) : myPath + "." + key;
            _this._childHash[key] = new Component(config, localPath);
        });
        return _this;
    }
    ObjectPart.prototype.process = function (data, rootList) {
        var _this = this;
        return _super.prototype.process.call(this, data, rootList).then(function (value) {
            if (value && ts_utils_1.isObject(value)) {
                var promises_1 = [];
                var result_1 = Object.create(null);
                Object.keys(_this._childHash).forEach(function (name) {
                    var promise = _this._childHash[name].process(data, rootList).then(function (itemValue) {
                        result_1[name] = itemValue;
                    });
                    promises_1.push(promise);
                });
                return Promise.all(promises_1).then(function () { return result_1; });
            }
            else {
                return value;
            }
        });
    };
    ObjectPart.prototype.getValue = function (data) {
        if (ts_utils_1.isObject(data)) {
            return data;
        }
        else {
            return null;
        }
    };
    return ObjectPart;
}(BasePart_1.BasePart));
exports.ObjectPart = ObjectPart;

},{"./BasePart":3,"ts-utils":undefined}],8:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Schema = (function () {
    function Schema(config) {
        var Component = config.type;
        this._children = new Component(config);
    }
    Schema.prototype.parse = function (data) {
        return this._children.process(data, []);
    };
    return Schema;
}());
exports.Schema = Schema;

},{}],9:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var BasePart_1 = require("./BasePart");
var config_1 = require("./config");
var ts_utils_1 = require("ts-utils");
var StringDatePart = (function (_super) {
    __extends(StringDatePart, _super);
    function StringDatePart(config, path) {
        var _this = _super.call(this, config, path) || this;
        _this.dateProcessor = ts_utils_1.date(_this.options.outPattern || config_1.OUT_DATE_PATTERN);
        return _this;
    }
    StringDatePart.prototype.getValue = function (data) {
        var date;
        if (data instanceof Date) {
            date = data;
        }
        if (typeof data === 'number') {
            date = new Date(data);
        }
        if (date) {
            return this.dateProcessor(date);
        }
        return null;
    };
    return StringDatePart;
}(BasePart_1.BasePart));
exports.StringDatePart = StringDatePart;

},{"./BasePart":3,"./config":11,"ts-utils":undefined}],10:[function(require,module,exports){
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var BasePart_1 = require("./BasePart");
var StringPart = (function (_super) {
    __extends(StringPart, _super);
    function StringPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StringPart.prototype.getValue = function (data) {
        switch (typeof data) {
            case 'string':
                return data;
            case 'number':
                return String(data);
            default:
                return null;
        }
    };
    return StringPart;
}(BasePart_1.BasePart));
exports.StringPart = StringPart;

},{"./BasePart":3}],11:[function(require,module,exports){
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OUT_DATE_PATTERN = 'DD.MM.YYYY';

},{}]},{},[1])(1)
});
