// @waves/oracle-data/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var STATUS_LIST;
(function (STATUS_LIST) {
    STATUS_LIST[STATUS_LIST["SCAM"] = -2] = "SCAM";
    STATUS_LIST[STATUS_LIST["SUSPICIOUS"] = -1] = "SUSPICIOUS";
    STATUS_LIST[STATUS_LIST["NOT_VERIFY"] = 0] = "NOT_VERIFY";
    STATUS_LIST[STATUS_LIST["DETAILED"] = 1] = "DETAILED";
    STATUS_LIST[STATUS_LIST["VERIFIED"] = 2] = "VERIFIED";
})(STATUS_LIST = exports.STATUS_LIST || (exports.STATUS_LIST = {}));
var DATA_PROVIDER_VERSIONS;
(function (DATA_PROVIDER_VERSIONS) {
    DATA_PROVIDER_VERSIONS[DATA_PROVIDER_VERSIONS["BETA"] = 0] = "BETA";
})(DATA_PROVIDER_VERSIONS = exports.DATA_PROVIDER_VERSIONS || (exports.DATA_PROVIDER_VERSIONS = {}));
var RESPONSE_STATUSES;
(function (RESPONSE_STATUSES) {
    RESPONSE_STATUSES["ERROR"] = "error";
    RESPONSE_STATUSES["OK"] = "ok";
    RESPONSE_STATUSES["EMPTY"] = "empty";
})(RESPONSE_STATUSES = exports.RESPONSE_STATUSES || (exports.RESPONSE_STATUSES = {}));
var DATA_ENTRY_TYPES;
(function (DATA_ENTRY_TYPES) {
    DATA_ENTRY_TYPES["INTEGER"] = "integer";
    DATA_ENTRY_TYPES["STRING"] = "string";
    DATA_ENTRY_TYPES["BINARY"] = "binary";
    DATA_ENTRY_TYPES["BOOLEAN"] = "boolean";
})(DATA_ENTRY_TYPES = exports.DATA_ENTRY_TYPES || (exports.DATA_ENTRY_TYPES = {}));
var DATA_PROVIDER_KEYS;
(function (DATA_PROVIDER_KEYS) {
    DATA_PROVIDER_KEYS["VERSION"] = "data_provider_version";
    DATA_PROVIDER_KEYS["NAME"] = "data_provider_name";
    DATA_PROVIDER_KEYS["LINK"] = "data_provider_link";
    DATA_PROVIDER_KEYS["EMAIL"] = "data_provider_email";
    DATA_PROVIDER_KEYS["LANG_LIST"] = "data_provider_lang_list";
    DATA_PROVIDER_KEYS["LOGO"] = "data_provider_logo";
})(DATA_PROVIDER_KEYS = exports.DATA_PROVIDER_KEYS || (exports.DATA_PROVIDER_KEYS = {}));
exports.DATA_PROVIDER_DESCRIPTION_PATTERN = 'data_provider_description_<LANG>';
var ORACLE_ASSET_FIELD_PATTERN;
(function (ORACLE_ASSET_FIELD_PATTERN) {
    ORACLE_ASSET_FIELD_PATTERN["VERSION"] = "version_<ASSET_ID>";
    ORACLE_ASSET_FIELD_PATTERN["STATUS"] = "status_<ASSET_ID>";
    ORACLE_ASSET_FIELD_PATTERN["LOGO"] = "logo_<ASSET_ID>";
    ORACLE_ASSET_FIELD_PATTERN["DESCRIPTION"] = "description_<LANG>_<ASSET_ID>";
    ORACLE_ASSET_FIELD_PATTERN["LINK"] = "link_<ASSET_ID>";
    ORACLE_ASSET_FIELD_PATTERN["TICKER"] = "ticker_<ASSET_ID>";
    ORACLE_ASSET_FIELD_PATTERN["EMAIL"] = "email_<ASSET_ID>";
})(ORACLE_ASSET_FIELD_PATTERN = exports.ORACLE_ASSET_FIELD_PATTERN || (exports.ORACLE_ASSET_FIELD_PATTERN = {}));
exports.PATTERNS = {
    ASSET_ID: '<ASSET_ID>',
    LANG: '<LANG>'
};
