// @waves/oracle-data/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var parse_1 = require("./parse");
var utils_1 = require("./utils");
var schemas_1 = require("./schemas");
function getProviderData(dataTxFields) {
    return parse_1.parseOracleData(utils_1.toHash('key')(dataTxFields));
}
exports.getProviderData = getProviderData;
function getProviderAssets(dataTxFields) {
    return parse_1.parseAssetData(utils_1.toHash('key')(dataTxFields));
}
exports.getProviderAssets = getProviderAssets;
function getFieldsFromData(data) {
    return schemas_1.DATA_TO_FIELDS.PROVIDER(data);
}
exports.getFieldsFromData = getFieldsFromData;
function getFieldsFromAsset(data) {
    return schemas_1.DATA_TO_FIELDS.ASSET(data);
}
exports.getFieldsFromAsset = getFieldsFromAsset;
function getFields(data) {
    if (utils_1.isProvider(data)) {
        return getFieldsFromData(data);
    }
    else {
        return getFieldsFromAsset(data);
    }
}
exports.getFields = getFields;
function getDifferenceByData(previous, next) {
    return utils_1.getFieldsDiff(toFields(previous), toFields(next));
}
exports.getDifferenceByData = getDifferenceByData;
function getDifferenceByFields(previous, next) {
    return utils_1.getFieldsDiff(previous, next);
}
exports.getDifferenceByFields = getDifferenceByFields;
function toFields(some) {
    return Array.isArray(some) ? some : getFields(some);
}
__export(require("./constants"));
