// @waves/signature-adapter/dist/adapters/Adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Signable_1 = require("../Signable");
var Adapter = (function () {
    function Adapter(networkCode) {
        networkCode = typeof networkCode === 'string' ? networkCode.charCodeAt(0) : networkCode;
        this.type = this.constructor.type;
        this._code = networkCode || Adapter._code || ('W').charCodeAt(0);
    }
    Adapter.prototype.makeSignable = function (forSign) {
        return new Signable_1.Signable(forSign, this);
    };
    Adapter.prototype.isAvailable = function () {
        return Promise.resolve();
    };
    Adapter.prototype.onDestroy = function (cb) {
        return;
    };
    Adapter.prototype.getNetworkByte = function () {
        return this._code || Adapter._code;
    };
    Adapter.initOptions = function (options) {
        Adapter._code = options.networkCode;
    };
    Adapter.getUserList = function () {
        return Promise.resolve([]);
    };
    Adapter.isAvailable = function () {
        return Promise.resolve(false);
    };
    Adapter.type = "seed";
    return Adapter;
}());
exports.Adapter = Adapter;
