// @waves/signature-adapter/dist/adapters/WavesKeeperAdapter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var _a;
var Adapter_1 = require("./Adapter");
var prepareTx_1 = require("../prepareTx");
var fieldValidator_1 = require("../prepareTx/fieldValidator");
var DEFAULT_TX_VERSIONS = (_a = {},
    _a[prepareTx_1.SIGN_TYPE.AUTH] = [1],
    _a[prepareTx_1.SIGN_TYPE.MATCHER_ORDERS] = [1],
    _a[prepareTx_1.SIGN_TYPE.CREATE_ORDER] = [1, 2, 3],
    _a[prepareTx_1.SIGN_TYPE.CANCEL_ORDER] = [1],
    _a[prepareTx_1.SIGN_TYPE.COINOMAT_CONFIRMATION] = [1],
    _a[prepareTx_1.SIGN_TYPE.ISSUE] = [2],
    _a[prepareTx_1.SIGN_TYPE.TRANSFER] = [2],
    _a[prepareTx_1.SIGN_TYPE.REISSUE] = [2],
    _a[prepareTx_1.SIGN_TYPE.BURN] = [2],
    _a[prepareTx_1.SIGN_TYPE.EXCHANGE] = [],
    _a[prepareTx_1.SIGN_TYPE.LEASE] = [2],
    _a[prepareTx_1.SIGN_TYPE.CANCEL_LEASING] = [2],
    _a[prepareTx_1.SIGN_TYPE.CREATE_ALIAS] = [2],
    _a[prepareTx_1.SIGN_TYPE.MASS_TRANSFER] = [1],
    _a[prepareTx_1.SIGN_TYPE.DATA] = [1],
    _a[prepareTx_1.SIGN_TYPE.SET_SCRIPT] = [1],
    _a[prepareTx_1.SIGN_TYPE.SPONSORSHIP] = [1],
    _a[prepareTx_1.SIGN_TYPE.SET_ASSET_SCRIPT] = [1],
    _a[prepareTx_1.SIGN_TYPE.SCRIPT_INVOCATION] = [1],
    _a);
var WavesKeeperAdapter = (function (_super) {
    __extends(WavesKeeperAdapter, _super);
    function WavesKeeperAdapter(_a, networkCode) {
        var address = _a.address, publicKey = _a.publicKey;
        var _this = _super.call(this, networkCode) || this;
        _this._onDestoryCb = [];
        _this._needDestroy = false;
        _this.handleUpdate = function (state) {
            if (!state.locked && (!state.account || state.account.address !== _this._address)) {
                _this._needDestroy = true;
                _this._onDestoryCb.forEach(function (cb) { return cb(); });
                _this._onDestoryCb = [];
                WavesKeeperAdapter.offUpdate(_this.handleUpdate);
            }
        };
        _this._address = address;
        _this._pKey = publicKey;
        WavesKeeperAdapter._initExtension();
        WavesKeeperAdapter.onUpdate(_this.handleUpdate);
        return _this;
    }
    WavesKeeperAdapter.prototype.isAvailable = function (ignoreLocked) {
        if (ignoreLocked === void 0) { ignoreLocked = false; }
        return __awaiter(this, void 0, void 0, function () {
            var data, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4, WavesKeeperAdapter.isAvailable(this.getNetworkByte())];
                    case 1:
                        _a.sent();
                        return [4, WavesKeeperAdapter._api.publicState()];
                    case 2:
                        data = _a.sent();
                        WavesKeeperAdapter._updateState(data);
                        if (data.locked) {
                            return [2, ignoreLocked ? Promise.resolve() : Promise.reject({ code: 4, msg: 'Keeper is locked' })];
                        }
                        if (data.account && data.account.address === this._address) {
                            return [2, Promise.resolve()];
                        }
                        return [3, 4];
                    case 3:
                        e_1 = _a.sent();
                        return [3, 4];
                    case 4: return [2, Promise.reject({ code: 5, msg: 'Keeper has another active account' })];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.isLocked = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, WavesKeeperAdapter.isAvailable()];
                    case 1:
                        _a.sent();
                        return [4, WavesKeeperAdapter._api.publicState()];
                    case 2:
                        data = _a.sent();
                        WavesKeeperAdapter._updateState(data);
                        if (data.locked) {
                            return [2, Promise.resolve()];
                        }
                        return [2];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.getSignVersions = function () {
        return WavesKeeperAdapter._txVersion;
    };
    WavesKeeperAdapter.prototype.onDestroy = function (cb) {
        if (this._needDestroy) {
            return cb();
        }
        this._onDestoryCb.push(cb);
    };
    WavesKeeperAdapter.prototype.getPublicKey = function () {
        return Promise.resolve(this._pKey);
    };
    WavesKeeperAdapter.prototype.getAddress = function () {
        return Promise.resolve(this._address);
    };
    WavesKeeperAdapter.prototype.getSeed = function () {
        return Promise.reject(Error('Method "getSeed" is not available!'));
    };
    WavesKeeperAdapter.prototype.signRequest = function (bytes, _, signData) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.isAvailable(true)];
                    case 1:
                        _a.sent();
                        return [4, WavesKeeperAdapter._api.signRequest(WavesKeeperAdapter._serializedData(signData))];
                    case 2: return [2, _a.sent()];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.signTransaction = function (bytes, amountPrecision, signData) {
        return __awaiter(this, void 0, void 0, function () {
            var dataStr, _a, proofs, signature;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4, this.isAvailable(true)];
                    case 1:
                        _b.sent();
                        return [4, WavesKeeperAdapter._api.signTransaction(WavesKeeperAdapter._serializedData(signData))];
                    case 2:
                        dataStr = _b.sent();
                        _a = JSON.parse(dataStr), proofs = _a.proofs, signature = _a.signature;
                        return [2, signature || proofs.pop()];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.signOrder = function (bytes, amountPrecision, signData) {
        return __awaiter(this, void 0, void 0, function () {
            var promise, dataStr, _a, proofs, signature;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4, this.isAvailable(true)];
                    case 1:
                        _b.sent();
                        switch (signData.type) {
                            case prepareTx_1.SIGN_TYPE.CREATE_ORDER:
                                promise = WavesKeeperAdapter._api.signOrder(WavesKeeperAdapter._serializedData(signData));
                                break;
                            case prepareTx_1.SIGN_TYPE.CANCEL_ORDER:
                                promise = WavesKeeperAdapter._api.signCancelOrder(WavesKeeperAdapter._serializedData(signData));
                                break;
                            default:
                                return [2, WavesKeeperAdapter._api.signRequest(WavesKeeperAdapter._serializedData(signData))];
                        }
                        return [4, promise];
                    case 2:
                        dataStr = _b.sent();
                        _a = JSON.parse(dataStr), proofs = _a.proofs, signature = _a.signature;
                        return [2, signature || proofs.pop()];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.signData = function (bytes) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.isAvailable(true)];
                    case 1:
                        _a.sent();
                        return [2, Promise.resolve('')];
                }
            });
        });
    };
    WavesKeeperAdapter.prototype.getPrivateKey = function () {
        return Promise.reject('No private key');
    };
    WavesKeeperAdapter.isAvailable = function (networkCode) {
        return __awaiter(this, void 0, void 0, function () {
            var error, data, e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, WavesKeeperAdapter._initExtension()];
                    case 1:
                        _a.sent();
                        if (!this._api) {
                            throw { code: 0, message: 'Install WavesKeeper' };
                        }
                        if (!(networkCode || Adapter_1.Adapter._code)) {
                            throw { code: 5, message: 'Set adapter network code' };
                        }
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, , 5]);
                        return [4, this._api.publicState()];
                    case 3:
                        data = _a.sent();
                        WavesKeeperAdapter._updateState(data);
                        if (data.txVersion) {
                            WavesKeeperAdapter._txVersion = data.txVersion;
                        }
                        return [3, 5];
                    case 4:
                        e_2 = _a.sent();
                        error = { code: 1, message: 'No permissions' };
                        return [3, 5];
                    case 5:
                        if (!error && data) {
                            if (!data.account) {
                                error = { code: 2, message: 'No accounts in waveskeeper' };
                            }
                            else if ((!data.account.address || !fieldValidator_1.isValidAddress(data.account.address, networkCode || Adapter_1.Adapter._code))) {
                                error = { code: 3, message: 'Selected network incorrect' };
                            }
                        }
                        if (error) {
                            throw error;
                        }
                        return [2, true];
                }
            });
        });
    };
    WavesKeeperAdapter.getUserList = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, WavesKeeperAdapter.isAvailable()];
                    case 1:
                        _a.sent();
                        return [2, WavesKeeperAdapter._api.publicState().then(function (data) {
                                WavesKeeperAdapter._updateState(data);
                                return [data.account];
                            })];
                }
            });
        });
    };
    WavesKeeperAdapter.initOptions = function (options) {
        Adapter_1.Adapter.initOptions(options);
        this.setApiExtension(options.extension);
        this._initExtension();
        try {
            this._api.publicState().then(WavesKeeperAdapter._updateState);
        }
        catch (e) {
        }
    };
    WavesKeeperAdapter.setApiExtension = function (extension) {
        var extensionCb;
        if (typeof extension === 'function') {
            extensionCb = extension;
        }
        else if (extension) {
            extensionCb = function () { return extension; };
        }
        WavesKeeperAdapter._getApiCb = extensionCb;
    };
    WavesKeeperAdapter.onUpdate = function (cb) {
        WavesKeeperAdapter._onUpdateCb.push(cb);
    };
    WavesKeeperAdapter.offUpdate = function (func) {
        WavesKeeperAdapter._onUpdateCb = WavesKeeperAdapter._onUpdateCb.filter(function (f) { return f !== func; });
    };
    WavesKeeperAdapter._updateState = function (state) {
        for (var _i = 0, _a = WavesKeeperAdapter._onUpdateCb; _i < _a.length; _i++) {
            var cb = _a[_i];
            cb(state);
        }
    };
    WavesKeeperAdapter._initExtension = function () {
        var _this = this;
        if (WavesKeeperAdapter._api || !WavesKeeperAdapter._getApiCb) {
            return WavesKeeperAdapter._api.initialPromise;
        }
        var wavesApi = WavesKeeperAdapter._getApiCb();
        if (wavesApi) {
            return wavesApi.initialPromise.then(function (api) {
                _this._api = api;
                _this._api.on('update', WavesKeeperAdapter._updateState);
                _this._api.publicState().then(function (state) {
                    if (state.txVersion) {
                        WavesKeeperAdapter._txVersion = state.txVersion;
                    }
                    WavesKeeperAdapter._updateState(state);
                });
            });
        }
    };
    WavesKeeperAdapter._serializedData = function (data) {
        return JSON.parse(JSON.stringify(data, function (key, value) { return value instanceof Uint8Array ? Array.from(value) : value; }));
    };
    WavesKeeperAdapter.type = "wavesKeeper";
    WavesKeeperAdapter._onUpdateCb = [];
    WavesKeeperAdapter._txVersion = DEFAULT_TX_VERSIONS;
    return WavesKeeperAdapter;
}(Adapter_1.Adapter));
exports.WavesKeeperAdapter = WavesKeeperAdapter;
